torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +752,134 @@ auto Engine::execute(const edge_list& roots,
+              << " is no longer valid, skipping execution";
+    return local_graph_task;
+  }
+
+  if (task->fn_ && !local_graph_task->has_error_.load()) {
+    AutoGradMode grad_mode(local_graph_task->grad_mode_);
+    try {
+      evaluate_function(local_graph_task, task->fn_.get(), task->inputs_);
+    } catch (std::exception& e) {
+      thread_on_exception(local_graph_task, task->fn_, e);
+    }
+  }
+
+  // Decrement the outstanding tasks.
+  --local_graph_task->outstanding_tasks_;
+
+  return local_graph_task;
+}
+
+void Engine::execute_until_ready_queue_empty(const std::shared_ptr<GraphTask>& graph_task) {
+  std::unique_lock<std::mutex> lock(print_mutex_);
+  std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
+  // LOG(ERROR) << "Engine::execute_until_ready_queue_empty on graphTask: " << graph_task;
+  while(!graph_task_rq->empty()) {
+    auto task = graph_task_rq->pop();
+    execute_node_task(task);
+    LOG(INFO)<<"executed task fn: " << task->fn_->name()<< " instance: "<< task->fn_
+      << ", for graph_task: " << graph_task << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+  // Check if we've completed execution.
+  if (graph_task_completed(graph_task)) {
+    // We don't need to explicitly notify the owner thread, since
+    // 'mark_graph_task_completed' would mark the Future as completed and this
+    // would notify the owner thread that the task has been completed.
+    mark_graph_task_completed(graph_task);
+    LOG(INFO) <<"finished graph_task: " << graph_task;
+  } else {
+    // schedule a continuation
+    at::launch([this, graph_task]() {
+        execute_until_ready_queue_empty(graph_task);
+    });
+    LOG(INFO) << "scheduled a continuation on graph_task: "  << graph_task
+      << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+
 }
 
 std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
     const std::shared_ptr<GraphTask>& graph_task,
-    std::shared_ptr<Node> graph_root) {
-  initialize_threads_pool();
+    std::shared_ptr<Node> graph_root,
+    bool async_mode) {
+  initialize_device_threads_pool();
   // Lock mutex for GraphTask.
   std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
-  ready_queue(at::kCPU).push(
+  ready_queue(graph_task, at::kCPU).push(
       NodeTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
-  // Not a worker
+  // worker_device == NO_DEVICE it's a CPU thread and it's trying to drive the
+  // autograd engine with corresponding GraphTask, and its NOT a re-entrant call
   if (worker_device == NO_DEVICE) {
     // graph_task_exec_post_processing is done when the Future is marked as
     // completed in mark_graph_task_completed.
-    return graph_task->future_result_;
+
+    // We set the worker_device to CPU_DEVICE only if worker_device was previously
+    // NO_DEVICE. Setting it to CPU afterwards allow us to detect whether this is
+    // a re-entrant call or not.
```

Comment:
This feels weird as you use two different mechanism to detect re-entrant calls: the worker_device and the thread local queue. Maybe only one is needed?

Reply:
Discussed offline, we should add an API to determine reentrant.